database/zookeeper.go

```diff
@@ -0,0 +1,156 @@
+	"github.com/samuel/go-zookeeper/zk"
+)
+
+type zkDriver struct {
+	connection types.ZkConnection
+}
+
+type zkConnector struct{}
+
+func (z zkConnector) Connect(zksStr string) (types.ZkConnection, error) {
+	zks := strings.Split(zksStr, ",")
+	conn, _, err := zk.Connect(zks, time.Second)
+
+	return conn, err
+}
+
+func createZKConnector() types.ZkConnectorInterface {
+	return types.ZkConnectorInterface(zkConnector{})
+}
+
+func encode(task *types.EremeticTask) ([]byte, error) {
+	encoded, err := json.Marshal(task)
+	return []byte(encoded), err
+}
+
+func createZKDriver(connector types.ZkConnectorInterface, path string) (TaskDB, error) {
+	if path == "" {
+		return nil, errors.New("Missing ZK path")
+	}
+
+	path = strings.Replace(path, "zk://", "", 1)
+
+	conn, err := connector.Connect(path)
+	if err != nil {
+		return nil, err
+	}
+
+	exists, _, err := conn.Exists("/eremetic")
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		flags := int32(0)
+		acl := zk.WorldACL(zk.PermAll)
+
+		_, err = conn.Create("/eremetic", nil, flags, acl)
+		if err != nil {
+			logrus.WithError(err).Error("Unable to create node.")
+			return nil, err
+		}
+	}
+
+	driver := zkDriver{connection: types.ZkConnection(conn)}
+
+	return driver, nil
+}
+
+func (z zkDriver) Close() {
+	z.connection.Close()
+}
+
+func (z zkDriver) Clean() error {
+	return z.connection.Delete("/eremetic/", -1)
+}
+
+func (z zkDriver) PutTask(task *types.EremeticTask) error {
+	path := fmt.Sprintf("/eremetic/%s", task.ID)
+
+	encode, err := encode(task)
+	if err != nil {
+		logrus.WithError(err).Error("Unable to encode task to byte-array.")
+		return err
+	}
+
+	exists, stat, err := z.connection.Exists(path)
+	if err != nil {
+		logrus.WithError(err).Error("Unable to check existance of database.")
+		return err
+	}
+
+	if exists {
+		_, err = z.connection.Set(path, encode, stat.Version)
+		return err
+	}
+
+	flags := int32(0)
+	acl := zk.WorldACL(zk.PermAll)
+	_, err = z.connection.Create(path, encode, flags, acl)
+	return err
+}
+
+func (z zkDriver) ReadTask(id string) (types.EremeticTask, error) {
+	task, err := z.ReadUnmaskedTask(id)
+
+	applyMask(&task)
+
+	return task, err
+}
+
+func (z zkDriver) ReadUnmaskedTask(id string) (types.EremeticTask, error) {
+	var task types.EremeticTask
+	path := fmt.Sprintf("/eremetic/%s", id)
+
+	bytes, _, err := z.connection.Get(path)
+	json.Unmarshal(bytes, &task)
+
+	return task, err
+
+}
+
+func (z zkDriver) ListNonTerminalTasks() ([]*types.EremeticTask, error) {
+	var tasks []*types.EremeticTask
+	paths, _, _ := z.connection.Children("/eremetic")
+	for _, p := range paths {
+		t, err := z.ReadTask(p)
+		if err != nil {
+			logrus.WithError(err).Error("Unable to read task from database, skipping")
+			continue
+		}
+		if !t.IsTerminated() {
+			applyMask(&t)
+			tasks = append(tasks, &t)
+		}
+	}
+
+	return tasks, nil
+}
+
+func (z zkDriver) Count() int {
+	var tasks []*types.EremeticTask
+
+	paths, _, _ := z.connection.Children("/eremetic")
```

Comment:
Same thing here with the full scan, but even worse because it's across the network. Found this but it looks a bit to involved for what we're doing, maybe we can just update a special /counter member?

http://www.mail-archive.com/zookeeper-user@hadoop.apache.org/msg01969.html